### PR TITLE
resource: add status RPC and flux-resource status command, misc cleanup

### PR DESCRIPTION
**Pull request by garlick**
This PR adds a `resource.status` RPC for querying R, excluded nodes, drained nodes (with timestamp and reason), online nodes, and offline nodes.  The idea is to support a `flux resource status` or similar command that utilizes the idset and hostlist python bindings just added by @grondo.

This is sort of a teaser WIP PR to see if this is the right concoction of the information.  Here's an example response:
```json
t/request/rpc resource.status </dev/null | jq
{
  "R": {
    "version": 1,
    "execution": {
      "R_lite": [
        {
          "rank": "0-15",
          "children": {
            "core": "0-3"
          }
        }
      ],
      "starttime": 0,
      "expiration": 0,
      "nodelist": [
        "jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro,jim-lemurpro"                       
      ]
    }
  },
  "drain": {
    "1-2": {
      "timestamp": 1605741116.0298321,
      "reason": "bad node!"
    },
    "15": {
      "timestamp": 1605741134.0681062,
      "reason": "down for service"
    }
  },
  "online": "0-15",
  "offline": "",
  "exclude": ""
}
```

In addition, since I added some general utility functions for working with objects that have idsets as keys, I took the opportunity to reduce XML as it is gathered, for the multiple-brokers-per-host case (It was too trivial to pass up).

Then a few bug fixes also.

The RPC doesn't have any tests yet - wanted to get some feedback first.

### CODE REVIEW

File: src/modules/resource/resource.c
Comment (chu11) [inline src/modules/resource/resource.c:175]: forgot a `goto error`?

Comment (garlick) [reply src/modules/resource/resource.c:175]: Doh! yup.

**Comment by garlick** [comment]
Forced a push with one line fix pointed out by @chu11 and also rebased on current master.

**Comment by grondo** [comment]
I've got a `flux resource status` tool built on top of this and it seems to working fine.

A new `flux resource status` utility can list the various target status, by default states with no nodes are excluded, `-v` and `-vv` give you more detail.

Additionally, `flux resource drain` with no args lists just the drained execution targets by timestamp.

```
ƒ(s=48,d=0,builddir) grondo@fluke25:~/git/f.git$ flux resource status
    STATUS NNODES RANKS           NODELIST
     avail     47 1-47            fluke[26-60,80-91]
     drain      1 0               fluke25
ƒ(s=48,d=0,builddir) grondo@fluke25:~/git/f.git$ flux resource status -v
    STATUS NNODES RANKS           NODELIST
     avail     47 1-47            fluke[26-60,80-91]
   offline      0                 
   exclude      0                 
     drain      1 0               fluke25
ƒ(s=48,d=0,builddir) grondo@fluke25:~/git/f.git$ flux resource status -vv
    STATUS NNODES RANKS           REASON                    NODELIST
     avail     47 1-47                                      fluke[26-60,80-91]
   offline      0                                           
   exclude      0                                           
     drain      1 0               no jobs on rank 0         fluke25
ƒ(s=48,d=0,builddir) grondo@fluke25:~/git/f.git$ flux resource drain
           TIMESTAMP RANK     REASON                         NODELIST
 2020-12-09T13:54:14 0        no jobs on rank 0              fluke25

```

My work is currently developed on top of this branch. Do we want to merge this branch first, or should I push my work here once I have developed some tests?

**Comment by garlick** [comment]
I can take off the WIP and we can merge this if it works for you.   Er, test coverage is not great in this PR but it probably will jump up again with yours.

**Comment by garlick** [comment]
Rebased on current master.

**Comment by grondo** [comment]
Ok, I've force-pushed the `flux resource status` implementation and update to `flux resource drain` here.

**Review by garlick — COMMENTED** [review]
This looks good to me @grondo.  I've poked at it manually and I like the new command!  Good addition.

Any idea what's up wit h the LGTM test?  It seems to be failing over in my latest PR too but I can't make heads or tails of the output.  It seems to get a "detached head" commit of flux and then give up?

Since this was a group effort maybe we'll need @chu11 to give a review and final approval?

**Comment by grondo** [comment]
> Any idea what's up wit h the LGTM test? It seems to be failing over in my latest PR too but I can't make heads or tails of the output. It seems to get a "detached head" commit of flux and then give up?

No, I don't understand what's going  on there but I assume it is a temporary LGTM issue. The "Python" build succeeded, and it checked out the project in "detached HEAD" mode, so I don't think that was the problem. Unfortunately there doesn't seem to be a way to trigger a new build of LGTM.

> Since this was a group effort maybe we'll need @chu11 to give a review and final approval?

I aggree. @chu11 should approve first.

**Comment by chu11** [comment]
did a quick skim and nothing stood out to me.  If you guys are happy, set MWP

**Review by chu11 — APPROVED** [review]

**Comment by grondo** [comment]
Sounds good, thanks @chu11!

**Pull request merged**
